File: app/src/main/java/com/arbaelbarca/listmovieskatalog/network/NetworkApi.java
package com.arbaelbarca.listmovieskatalog.network;

import com.arbaelbarca.listmovieskatalog.utils.Constants;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class NetworkApi {
    private ApiServices apiServices;
    private static NetworkApi networkApi;
    private String credentials;

    private NetworkApi() {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(Constants.BASE_URL_MOVIES)
                .client(provideOkHttpClient())
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .build();

        apiServices = retrofit.create(ApiServices.class);

    }



    private static OkHttpClient provideOkHttpClient() {
        return new OkHttpClient.Builder()
                .addInterceptor(getLoggingInterceptor())
                .addInterceptor(chain -> {
                    Request request = chain.request();
                    Request authenticatedRequest = request.newBuilder()
                            .header("Authorization", "Bearer eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiJkY2Y1NzYzNjkxYTJhOTYwZDJkMjQyMDk3Y2RiMWY2YyIsInN1YiI6IjU5ODU1NDEzOTI1MTQxNDM4NzAxOGMyZSIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.Pg0vpPU8iOhVHW3g7ZUEOkA3cMp9JEwMpx9AhD-92Tk").build();
                    return chain.proceed(authenticatedRequest);
                })
                .readTimeout(15, TimeUnit.SECONDS)
                .connectTimeout(15, TimeUnit.SECONDS)
                .writeTimeout(15, TimeUnit.SECONDS)
                .build();
    }

    private static HttpLoggingInterceptor getLoggingInterceptor() {
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        return interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
    }

    public ApiServices getAPI() {
        return apiServices;
    }

    public static NetworkApi getInstance() {
        if (networkApi == null)
            networkApi = new NetworkApi();
        return networkApi;
    }
}

File: app/src/main/java/com/robytakwa/listpicturekatalog/network/NetworkApi.kt
package com.robytakwa.listpicturekatalog.network

import com.robytakwa.listpicturekatalog.utils.Constants
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

class NetworkApi private constructor() {
    val api: ApiServices
    private val credentials: String? = null

    init {
        val retrofit = Retrofit.Builder()
                .baseUrl(Constants.BASE_URL_MOVIES)
                .client(provideOkHttpClient())
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .build()

        api = retrofit.create(ApiServices::class.java)

    }

    companion object {
        private var networkApi: NetworkApi? = null


        private fun provideOkHttpClient(): OkHttpClient {
            return OkHttpClient.Builder()
                    .addInterceptor(loggingInterceptor)
                    .addInterceptor { chain ->
                        val request = chain.request()
                        val authenticatedRequest = request.newBuilder()
                                .header("Authorization", "Bearer eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiJkY2Y1NzYzNjkxYTJhOTYwZDJkMjQyMDk3Y2RiMWY2YyIsInN1YiI6IjU5ODU1NDEzOTI1MTQxNDM4NzAxOGMyZSIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.Pg0vpPU8iOhVHW3g7ZUEOkA3cMp9JEwMpx9AhD-92Tk").build()
                        chain.proceed(authenticatedRequest)
                    }
                    .readTimeout(15, TimeUnit.SECONDS)
                    .connectTimeout(15, TimeUnit.SECONDS)
                    .writeTimeout(15, TimeUnit.SECONDS)
                    .build()
        }

        private val loggingInterceptor: HttpLoggingInterceptor
            get() {
                val interceptor = HttpLoggingInterceptor()
                return interceptor.setLevel(HttpLoggingInterceptor.Level.BODY)
            }

        val instance: NetworkApi
            get() {
                if (networkApi == null)
                    networkApi = NetworkApi()
                return networkApi!!
            }
    }
}
